### ISSUE TITLE
Eval bug:Phi-3 / Phi-4 LongRoPE: short sequences silently use long-context RoPE factors when the allocated context exceeds original_max_position_embeddings

### ISSUE THREAD

**Issue by b-bargabus** [ggml-org/llama.cpp#24823]
### Name and Version

llama-server version: 9608 (70b54e140) - the code driving this (get_rope_factors keying on n_ctx_seq in llama-model.cpp) is unchanged in current master, 9728 (fabde3b)

(same thing will happen in llama-cli, it's just easier to test in llama-server)

### Operating systems

Linux

### GGML backends

CPU, CUDA

### Hardware

It's independent of hardware, confirmed on multiple and the code cited below affects all.

### Models

phi3/4 family. It shows up most drastically in fine tunes but the bug is there for all of them, the default model just seems to be less affected by the issue but is equally wrong.

### Problem description & steps to reproduce

For LongRoPE models (Phi-3 / Phi-4 family), llama.cpp selects the long vs short RoPE factor set based on the allocated context size (n_ctx_seq), not the actual sequence length. As a result, when a model is loaded with a context larger than original_max_position_embeddings (4096 for Phi-4-mini), every sequence, including very short ones, is encoded with the long-context factors. The reference implementation (Hugging Face transformers) selects the factor set from the current sequence length, so short prompts get the short factors.

The divergence is silent, and its impact is model-dependent. Some models tolerate the mis-scaled encoding with no measurable change in quality, but it can be catastrophic for others, including a legitimate fine-tune that is correct at a smaller context and under the reference implementation. Because the impact varies by model, the failure is easily misattributed to the model or the fine-tuning process rather than to the serving path.

Environment

llama.cpp version: version: 9608 (70b54e140), reproduced with llama-server on CPU (-ngl 0), so the result is backend-independent. The faulty code is unchanged in current master (9728 / fabde3b); code references below.
Model: a Phi-4-mini-instruct fine-tune (Phi3 architecture), LongRoPE, rope.dimension_count = 96, rope.freq_base = 10000, rope.scaling.original_context_length = 4096, training context 131072. Served as an f16 GGUF with the LongRoPE factor tensors present, so this is not a quantization or missing-metadata issue.
Reference implementation: transformers 5.5.0.

Steps to reproduce

Minimal reproduction, native llama-server, CPU only, identical in every respect except the allocated context, which differs only by crossing the 4096 boundary:

```
# correct: allocation within the 4096 original_max -> short RoPE factors
llama-server -m phi4-mini-ft.f16.gguf -ngl 0 -c 4096 --host 127.0.0.1 --port 8099

# broken: same model, same prompt, same flags, only -c crosses 4096 -> long factors
llama-server -m phi4-mini-ft.f16.gguf -ngl 0 -c 8192 --host 127.0.0.1 --port 8099

# one greedy request in each case; the prompt is a single document of about 640 tokens:
curl -s localhost:8099/completion \
  -d '{"prompt":"<wrapped prompt>","n_predict":1024,"temperature":0,"cache_prompt":false}'
```

The server logs confirm the context took: n_ctx_seq (4096) and n_ctx_seq (8192) respectively, both well below the 131072 training context.

At -c 4096 the model returned a complete, correct extraction (field F1 = 1.000):

```
{"products": [{"product_id": "p1", "product_number": "WA-SL150", "product_name": "OmniLift Stacker SL-150"}],
 "options": [{"option_id": "o1", "product_id": "p1", "option_name": "Lift capacity"}, ...],
 "option_values": [{"value_id": "v1", "option_id": "o1", "value_name": "1000 kg"}, ...],
 "rules": [{"option_id": "o1", "min": 1, "max": 1}, ...]}
```

At -c 8192, with the same prompt, the model stopped grounding in the document after the first record and reproduces the output-format template with literal placeholder tokens (field F1 = 0.000):

```
{"products": [{"product_id": "p1", "product_number": "WA-SL150", "product_name": "OmniLift Stacker SL-150"}],
 "options": [{"option_id": "o1", "product_name": "option_name"}, {"option_id": "o2", "product_name": "option_name"}, ...],
 "option_values": [{"value_id": "v1", "option_id": "<ref>", "value_name": "value_name"}, ...],
 "rules": [{"option_id": "<ref>", "min": <int>, "max": <int or null>}, ...]}
```

The literal <ref>, value_name, <int or null>, and ... are placeholder tokens from the prompt's format description, not values read from the document. The only thing that changed between the two runs is -c.

Supporting measurements

I confirmed the issue direct via llama-server to rule out ollama (plus found the specific error in code) but I had first tested in ollama. Across the eight short documents that failed, served via Ollama 0.30.6 (bundled llama.cpp), greedy:

```
num_ctx = 4096    mean F1 = 0.992
num_ctx = 131072  mean F1 = 0.105
```

The same fine-tuned weights served through transformers 5.5.0 (which selects the short factors for these sub-4096 sequences) score about 0.94 on the full 105-document test set, matching the num_ctx <= 4096 llama.cpp result and confirming the weights are correct.

As a control, the base Phi-4-mini converted through the same pipeline is essentially unaffected by the allocation:

```
base    num_ctx = 4096    mean F1 = 0.913
base    num_ctx = 131072  mean F1 = 0.919
```

So the wrong factor selection is not catastrophic for every model. It is catastrophic for one that depends more heavily on precise positional encoding, which is exactly the kind of model that is correct at num_ctx <= 4096 and under the reference implementation. That model-dependence is what makes the issue easy to misattribute to the model rather than the serving path. tl;dr The issue is real but doesn't show up to a significant degree with the reference phi4 model, any finetune amplifies the existing bug though and becomes unusable quickly.

Expected behavior

A sequence whose length is at or below original_max_position_embeddings should be encoded with the short factor set, regardless of how much context capacity was allocated. This matches the LongRoPE reference and the transformers implementation.

In transformers (modeling_rope_utils._compute_longrope_parameters), long_factor is used only when seq_len > original_max_position_embeddings; otherwise short_factor is used.

Actual behavior

llama.cpp chooses the factor set from the allocated context, not the sequence length.

src/llama-model.cpp (get_rope_factors):

```cpp
ggml_tensor * llama_model::get_rope_factors(const llama_cparams & cparams, int il) const {
    const uint32_t n_ctx_seq = cparams.n_ctx_seq;

    // choose long/short freq factors based on the context size
    if (layers[il].rope_freqs != nullptr) {
        return layers[il].rope_freqs;
    }

    if (n_ctx_seq > hparams.n_ctx_orig_yarn) {
        return layers[il].rope_long;
    }

    return layers[il].rope_short;
}
```

n_ctx_seq is the allocated per-sequence context, set in src/llama-context.cpp:

```cpp
cparams.n_ctx_seq = cparams.n_ctx;                  // single sequence
// or
cparams.n_ctx_seq = cparams.n_ctx / cparams.n_seq_max;
```

So whenever n_ctx_seq > n_ctx_orig_yarn (4096), the long factors are returned for all positions, including a 700-token prompt that should use the short factors. For Phi-4-mini the long factors scale the inverse frequencies by up to roughly 47x on the low-frequency dimensions, which materially changes the positional encoding relative to the short factors the model (and the reference implementation) expect for short sequences.

Why this is easy to miss

The failure is silent. There is no warning, the JSON or text output is still well-formed, and for some inputs it is even correct, so the degradation looks like a model-quality or fine-tuning problem rather than a serving artifact. It only appears when the allocated context exceeds original_max_position_embeddings, which is common because tools default to or allow large contexts for a model whose training context is 131072.

Proposed fix

Choose the factor set from the actual sequence length (or the maximum position in the batch) rather than from the static allocation. This is what the reference implementation does, and for any sequence that stays within n_ctx_orig_yarn it is unambiguous and needs no special machinery: the short factors are simply correct, whatever the allocation. The current code returns the long factors for these sequences only because n_ctx_seq reflects the allocation rather than the length.

The one genuinely subtle case is a single sequence that crosses n_ctx_orig_yarn during generation while a KV cache is populated, since keys cached under one factor set cannot be cheaply re-roped to the other. That edge case is shared by any cached-attention implementation, including the reference, and it does not block fixing the common case where the whole sequence is at or below n_ctx_orig_yarn.

As a minimum mitigation, emit a warning when a LongRoPE model is loaded with n_ctx_seq > n_ctx_orig_yarn, so the silent degradation becomes visible.

Possible related issue

ggml-org/llama.cpp issue 12091 discusses Phi-4 RoPE handling. Worth checking for overlap or duplication before filing.

### First Bad Commit

not bisected, present in 9608 and unchanged in current master (9728). long standing, not a regression.

### Relevant log output

<details>
<summary>Logs</summary>

```console

```
</details>

**Comment by b-bargabus** [comment]
Thank you!
